hail/src/main/scala/is/hail/expr/ir/Binds.scala

```diff
@@ -3,321 +3,411 @@ package is.hail.expr.ir
-      case RelationalLetTable(name, value, _) => if (i == 1) FastSeq(name -> value.typ) else empty
-      case RelationalLetMatrixTable(name, value, _) =>
-        if (i == 1) FastSeq(name -> value.typ) else empty
-      case RelationalLetBlockMatrix(name, value, _) =>
-        if (i == 1) FastSeq(name -> value.typ) else empty
-      case _ => empty
+  /** Like [[Bindings.apply]], but keeps separate any new bindings introduced by `ir`. Always
+    * satisfies the identity
+    * {{{
+    * Bindings.segregated(ir, i, baseEnv).unified == Bindings(ir, i, baseEnv)
+    * }}}
+    */
+  def segregated[A](ir: BaseIR, i: Int, baseEnv: BindingEnv[A]): SegregatedBindingEnv[A, Type] =
+    apply(ir, i, SegregatedBindingEnv(baseEnv))
+
+  private def childEnvMatrix[E <: GenericBindingEnv[E, Type]](ir: MatrixIR, i: Int, _baseEnv: E)
+    : E = {
+    val baseEnv = _baseEnv.onlyRelational()
+    ir match {
+      case MatrixMapRows(child, _) if i == 1 =>
+        baseEnv
+          .createAgg.createScan
+          .bindEval(child.typ.rowBindings: _*)
+          .bindEval("n_cols" -> TInt32)
+          .bindAgg(child.typ.entryBindings: _*)
+          .bindScan(child.typ.rowBindings: _*)
+      case MatrixFilterRows(child, _) if i == 1 =>
+        baseEnv.bindEval(child.typ.rowBindings: _*)
+      case MatrixMapCols(child, _, _) if i == 1 =>
+        baseEnv
+          .createAgg.createScan
+          .bindEval(child.typ.colBindings: _*)
+          .bindEval("n_rows" -> TInt64)
+          .bindAgg(child.typ.entryBindings: _*)
+          .bindScan(child.typ.colBindings: _*)
+      case MatrixFilterCols(child, _) if i == 1 =>
+        baseEnv.bindEval(child.typ.colBindings: _*)
+      case MatrixMapEntries(child, _) if i == 1 =>
+        baseEnv.bindEval(child.typ.entryBindings: _*)
+      case MatrixFilterEntries(child, _) if i == 1 =>
+        baseEnv.bindEval(child.typ.entryBindings: _*)
+      case MatrixMapGlobals(child, _) if i == 1 =>
+        baseEnv.bindEval(child.typ.globalBindings: _*)
+      case MatrixAggregateColsByKey(child, _, _) =>
+        if (i == 1)
+          baseEnv
+            .bindEval(child.typ.rowBindings: _*)
+            .createAgg.bindAgg(child.typ.entryBindings: _*)
+        else if (i == 2)
+          baseEnv
+            .bindEval(child.typ.globalBindings: _*)
+            .createAgg.bindAgg(child.typ.colBindings: _*)
+        else baseEnv
+      case MatrixAggregateRowsByKey(child, _, _) =>
+        if (i == 1)
+          baseEnv
+            .bindEval(child.typ.colBindings: _*)
+            .createAgg.bindAgg(child.typ.entryBindings: _*)
+        else if (i == 2)
+          baseEnv
+            .bindEval(child.typ.globalBindings: _*)
+            .createAgg.bindAgg(child.typ.rowBindings: _*)
+        else baseEnv
+      case RelationalLetMatrixTable(name, value, _) if i == 1 =>
+        baseEnv.bindRelational(name -> value.typ)
+      case _ =>
+        baseEnv
```

Comment:
> I'd prefer to see an exhaustive match rather than a default case because if we add a new node we'll get an immediate failure.

I get that, but I disagree in this case. A compiler has to track lots of properties of IR nodes, like the binding structure here, nesting depth, existence of side effects, strictness re. missingness, whether it is a statically computable constant, etc. In each case there is a default which the majority of nodes satisfy, and only a subset needs to override the default. I don't think forcing the code to explicitly define every property for every node has enough benefit to be worth drowning out the meaningful signal in a sea of boilerplate.

What I would really like to do is move to the other side of the expression problem, with these properties defined directly on the IR class, with the default specified through inheritence. That way all the semantically meaningful properties of a node are in one place. This doesn't really change anything--you still need to remember to override the default when appropriate--but I think having everything in one place, and being able to look at a list of overridable methods on the parent, would make it harder to forget.

Reply:
Also, we don't add new nodes very often. I'd rather optimize for reading rather than writing.

Reply:
> drowning out the meaningful signal in a sea of boilerplate.

A little dramatic perhaps haha. I understand your point about not adding nodes often, but from my own experience of adding IR nodes we do get weird and wonderful errors from not not overriding defaults. 
Having a case with an exhaustive list of nodes with default implementations is not a significant amount of boilerplate.
I'm of the mind that all these defaults actually makes it harder to work on this codebase. We have to resort to dev docs to list all the steps we need to do to change the code. We have a compiler, we should use it.

> default specified through inheritence

Please no. Inheritance is one of the largest sources of complexity in this codebase, especially when it comes to mixing state and behaviour. This is like death by a thousand cuts for me (my turn for dramatics :) )